drivers/spi/spi_rpi_pico_pio.c

```diff
@@ -622,33 +545,33 @@ static void spi_pico_pio_txrx_3_wire(const struct device *dev)
 
 				switch (data->dfs) {
 				case 4: {
-					txrx = ((uint32_t *)txbuf)[data->tx_count];
+					txrx = sys_get_be32(txbuf + (data->tx_count * 4));
 					spi_pico_pio_sm_put32(data->pio, data->pio_sm, txrx);
-					data->tx_count += 4;
 				} break;
 
 				case 2: {
-					txrx = ((uint16_t *)txbuf)[data->tx_count];
+					txrx = sys_get_be16(txbuf + (data->tx_count * 2));
 					spi_pico_pio_sm_put16(data->pio, data->pio_sm, txrx);
-					data->tx_count += 2;
 				} break;
 
 				case 1: {
 					txrx = ((uint8_t *)txbuf)[data->tx_count];
 					spi_pico_pio_sm_put8(data->pio, data->pio_sm, txrx);
-					data->tx_count++;
 				} break;
 
 				default:
 					LOG_ERR("Support fot %d bits not enabled", (data->dfs * 8));
 					break;
 				}
+				data->tx_count++;
 			}
 		}
-		while ((!pio_sm_is_tx_fifo_empty(data->pio, data->pio_sm)) ||
-		       (!spi_pico_pio_sm_complete(data))) {
+		/* Wait for the state machine to complete the cycle */
+		/* before resetting the PIO for reading.            */
+		do {
 			k_sleep(K_TICKS(data->tx_period_ticks));
```

Comment:
Looking at other drivers it seems that waiting on a bit without sleeping between iterations is legitimate. If any higher priority task needs to run it will get its runtime.

Reply:
I'm sorry, but I'm not following you.  Are you suggesting using a timer instead of k_sleep()?  Or perhaps I didn't explain the problem clearly:  Even after the last word has been shifted from the FIFO, clocked out by the PIO program, and the PIO program counter has returned to the starting point, a delay is still needed to allow the SPI bus signals to complete before the CPU can reset the PIO state machine to start the receive cycle.  I can't think of any way for the CPU to detect the state of the SPI bus, which is why I added the delay here.

I did try to minimize the needed delay, though perhaps waiting the full word size is excessive.  I'll test this with just one SPI clock cycle and see if that's enough.

Reply:
Only one SPI bit time is needed;  I tested it at both the fast end and the slow end of the BME280 clock range.  (The issue is more of a problem at slow SPI clock rates.)

Reply:
@ThreeEights what worries me is that putting delays in drivers may impact throughput. I once debugged a slow UART interface just to discover there was a similar delay adding a few milliseconds between each buffer sent.
Although throughput might not be a big issue here, I still think it's best not to have delays in driver code. The kernel's delay resolution is in ticks, and a tick is usually 1 or 0.5 ms. Since SPI works in the MHz range, the needed delay for one bit is in the microseconds range. My suggestion is to busy wait that delay instead of sleeping through it:
```
while ((!pio_sm_is_tx_fifo_empty(data->pio, data->pio_sm)) ||
			 (!spi_pico_pio_sm_complete(data)));
```

Reply:
@yonsch - I agree with everything you're saying here, but it's obvious I have not yet communicated what, exactly, is the problem.  The wait loop you show is the original code I had, with no sleep delay, which fails:
* The TX FIFO will be empty when the last data is transferred to the output shift register, before any bits have been sent.  The call to pio_sm_is_tx_fifo_empty() just ensures that all the queued data has been passed to the state machine.
* spi_pico_pio_sm_complete() compares the state machine's PC with the start address of the send program.  That tells the CPU that the SM has returned to the starting point, but the last clock transition has not yet propagated over the CLK line.
* When this loop completes, the next step is resetting the state machine to turn the data line around for receiving the response.
If the state machine is reset before the clock transition finishes, the data and clock lines are now out of sync and the next few exchanges are garbage.  I see this issue showing up with a BME280 sensor and an external CYW4343W WiFi module as well as the CYW43439 on the Pico W.
I'm open to suggestions for alternatives, but I've been unable to figure out any way for the CPU to detect that the clock signal has propagated to the peripheral.  Absent any other solution, I think the k_sleep() call in this case is justified:
* This only applies when running in 3-wire mode.
* The delay takes place after all the send data in this SPI transaction has been transmitted.
* For the RPI Pico and Pico W, CONFIG_SYS_CLOCK_TICKS_PER_SEC is 10000, which implies a tick time of 0.1 msec.
* For most of the SPI clock speed range, waiting for one "one bit time" amounts to 1 tick, or 100 microseconds.
Again, if you can think of any other way to let the SM finish without inserting this k_sleep(), please let me know.

Reply:
Okay, this satisfies me, so I'll let @soburi decide

Reply:
@yonsch @soburi This is a bit embarrassing, but it turns out I can move the necessary delay into the PIO program and remove the k_sleep() completely.  Once I can get all the compliance checks to pass, this should be a much cleaner solution.
Taking out the k_sleep(), the delay between the Send and Receive stages goes from an average of 265 microseconds to 170, with some cycles being substantially faster.

Reply:
@yonsch @soburi - Finally cleared up all the compliance and infrastructure update issues.  Can you please take another look at the revised code?